src/serac/physics/utilities/variational_form/tensor.hpp

```diff
@@ -0,0 +1,1267 @@
+        AB[i][j] = AB[i][j] + A[i][k] * B[k][j];
+      }
+    }
+  }
+  return AB;
+}
+
+template <typename S, typename T, int m, int n>
+constexpr auto dot(const tensor<S, m>& A, const tensor<T, m, n>& B)
+{
+  tensor<decltype(S{} * T{}), n> AB{};
+  for (int i = 0; i < n; i++) {
+    for (int j = 0; j < m; j++) {
+      AB[i] = AB[i] + A[j] * B[j][i];
+    }
+  }
+  return AB;
+}
+
+template <typename S, typename T, int m, int n>
+constexpr auto dot(const tensor<S, m, n>& A, const tensor<T, n>& B)
+{
+  tensor<decltype(S{} * T{}), m> AB{};
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      AB[i] = AB[i] + A[i][j] * B[j];
+    }
+  }
+  return AB;
+}
+
+template <typename S, typename T, int n>
+constexpr auto dot(const tensor<S, n>& A, const tensor<T, n>& B)
+{
+  decltype(S{} * T{}) AB{};
+  for (int i = 0; i < n; i++) {
+    AB += A[i] * B[i];
+  }
+  return AB;
+}
+
+template <typename S, typename T, int m, int n, int p>
+constexpr auto dot(const tensor<S, m, n, p>& A, const tensor<T, p>& B)
+{
+  tensor<decltype(S{} * T{}), m, n> AB{};
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      for (int k = 0; k < p; k++) {
+        AB[i][j] += A[i][j][k] * B[k];
+      }
+    }
+  }
+  return AB;
+}
+
+template <typename S, typename T, typename U, int m, int n>
+constexpr auto dot(const tensor<S, m>& u, const tensor<T, m, n>& A, const tensor<U, n>& v)
+{
+  decltype(S{} * T{} * U{}) uAv{};
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      uAv += u[i] * A[i][j] * v[j];
+    }
+  }
+  return uAv;
+}
+
+template <typename S, typename T, int m, int n, int p, int q>
+constexpr auto ddot(const tensor<S, m, n, p, q>& A, const tensor<T, p, q>& B)
+{
+  tensor<decltype(S{} * T{}), m, n> AB{};
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      for (int k = 0; k < p; k++) {
+        for (int l = 0; l < q; l++) {
+          AB[i][j] += A[i][j][k][l] * B[k][l];
+        }
+      }
+    }
+  }
+  return AB;
+}
+
+template <typename S, typename T, int m, int n, int p>
+constexpr auto ddot(const tensor<S, m, n, p>& A, const tensor<T, n, p>& B)
+{
+  tensor<decltype(S{} * T{}), m> AB{};
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      for (int k = 0; k < p; k++) {
+        AB[i] += A[i][j][k] * B[j][k];
+      }
+    }
+  }
+  return AB;
+}
+
+template <typename S, typename T, int m, int n>
+constexpr auto ddot(const tensor<S, m, n>& A, const tensor<T, m, n>& B)
+{
+  decltype(S{} * T{}) AB{};
+  for (int i = 0; i < m; i++) {
+    for (int j = 0; j < n; j++) {
+      AB += A[i][j] * B[i][j];
+    }
+  }
+  return AB;
+}
+
+template <typename S, typename T, int m, int n, int p>
+constexpr auto operator*(const tensor<S, m, n>& A, const tensor<T, n, p>& B)
+{
+  return dot(A, B);
+}
+
+template <typename S, typename T, int m, int n>
+constexpr auto operator*(const tensor<S, m>& A, const tensor<T, m, n>& B)
+{
+  return dot(A, B);
+}
+
```

Comment:
I just added this in my branch, but I think it'll probably be used at some point.
```suggestion
template <typename S, typename T, int m>
constexpr auto operator*(const tensor<S, m>& A, const tensor<T, m>& B)
{
  return dot(A, B);
}
```